libsolidity/formal/CHC.cpp

```diff
@@ -677,20 +677,31 @@ void CHC::externalFunctionCall(FunctionCall const& _funCall)
 	bool usesStaticCall = kind == FunctionType::Kind::BareStaticCall ||
 		function->stateMutability() == StateMutability::Pure ||
 		function->stateMutability() == StateMutability::View;
+
 	if (!usesStaticCall)
 	{
 		state().newState();
 		for (auto const* var: m_stateVariables)
 			m_context.variable(*var)->increaseIndex();
 	}
 
-	auto postCallState = vector<smtutil::Expression>{state().state()} + currentStateVariables();
 	auto error = errorFlag().increaseIndex();
+
+	Predicate const& callPredicate = *createSymbolicBlock(
+		nondetInterfaceSort(*m_currentContract, state()),
+		"nondet_call_" + uniquePrefix(),
+		PredicateType::ExternalCallUntrusted,
+		&_funCall
+	);
+	auto postCallState = vector<smtutil::Expression>{state().state()} + currentStateVariables();
 	vector<smtutil::Expression> stateExprs{error, state().thisAddress(), state().abi(), state().crypto()};
+
 	auto nondet = (*m_nondetInterfaces.at(m_currentContract))(stateExprs + preCallState + postCallState);
-	// TODO this could instead add the summary of the called function, where that summary
-	// basically has the nondet interface of this summary as a constraint.
-	m_context.addAssertion(nondet);
+	auto nondetCall = callPredicate(stateExprs + preCallState + postCallState);
+
+	addRule(smtutil::Expression::implies(nondet, nondetCall), nondetCall.name);
```

Comment:
The idea here is similar to the internal calls: in order to know which external call led to each reentrant call, for each external call we add
- a new `nondet_call_id` predicate
- the rule `nondet(s, s') => nondet_call_id(s, s')`
and we use `nondet_call_id` as the "call" constraint. This is also the predicate we look for in the counterexample graph.